Managed/Autonomous agents lose unsent/retriable events when principal connection is lost

**Describe the bug**

The following is the current behaviour:
1) Agent establishes initial connection to principal
-  _(... additional time passes ...)_
-  ( _agent loses connection to principal_ )
2) `maintainConnection` (`agent/connection.go`) creates a new connection to principal (once connection to principal becomes possible)
3) `handleStreamEvents` (`agent/connection.go`) creates a `EventWriter` to fully replace the old `EventWriter`, and that new EventWriter begins receiving events

However, 3) is incorrect, as the old `EventWriter` object still exists from 1), and still has events left to send in `unsentEvents` and `sentEvents` (though sending events will fail because that old `EventWriter` is still trying to use the previous dead connection)

Principal code does not have this problem, because it calls `EventWriter.UpdateTarget` when a new connection is established (thus preserving the existing events).

Found using chaos engineering via toxiproxy, which I will contribute separately.

Reproducible on `main` branch as commit `31282236c1d088ac579544f71dab8720a90ea8c7` (Jan 14).


### To reproduce

This will simulate persistent failures from managed agent <-> principal. (Autonomous is not modified here)
```
# Pull branch for bug reproduction
git clone https://github.com/jgwest/argocd-agent
git checkout delme-repro-argocd-agent-715-jan-2026

# setup vcluster and enabled argocd-agent config 
make setup-e2e

# start toxiproxy server (or docker)
podman run --rm --net=host -it ghcr.io/shopify/toxiproxy

# chaos tester telsl toxiproxy server to block all connections between principal<->agent every 10-15 seconds, for 5-10 seconds. 
go run hack/chaos-tester/main.go

# Start local argocd-agent (agent uses toxiproxy as intermediary to principal)
ARGOCD_AGENT_REMOTE_PORT=8475 make start-e2e

# Run Test_SyncManaged over and over until it fails
# For me, it takes 20-25 minutes to fail.
until-fail.sh   go test -count=1 -v -v -run TestSyncTestSuite/Test_SyncManaged ./test/e2e/

# until-fail.sh is simple shell script:
# https://gist.github.com/jgwest/7048a765d398519837f990120cf3fdd0
```

### Example failure logs when running repro steps

```

# Round N-1 passes
=== RUN   TestSyncTestSuite/Test_SyncManaged
    fixture.go:103: Test begun at: 2026-01-29 12:50:55.562902512 -0500 EST m=+0.108581668
    sync_test.go:104: jgw: waiting for outofsync guestbook agent-managed 2026-01-29 12:51:02.58232843 -0500 EST m=+7.128007588
    sync_test.go:113: jgw: starting syncapplication of guestbook agent-managed 2026-01-29 12:51:03.587190644 -0500 EST m=+8.132869811
    sync_test.go:118: jgw: completing syncapplication 2026-01-29 12:51:03.601917405 -0500 EST m=+8.147596602
    sync_test.go:127: jgw: completing eventually 2026-01-29 12:51:04.607752216 -0500 EST m=+9.153431369
    fixture.go:107: Test ended at: 2026-01-29 12:51:06.64945961 -0500 EST m=+11.195138776
--- PASS: TestSyncTestSuite (28.34s)
    --- PASS: TestSyncTestSuite/Test_SyncManaged (28.30s)
PASS
ok  	github.com/argoproj-labs/argocd-agent/test/e2e	28.373s
=== RUN   TestSyncTestSuite

# Round N fails
=== RUN   TestSyncTestSuite/Test_SyncManaged
    fixture.go:103: Test begun at: 2026-01-29 12:51:26.843821979 -0500 EST m=+0.110578000
    sync_test.go:104: jgw: waiting for outofsync guestbook agent-managed 2026-01-29 12:51:27.862541937 -0500 EST m=+1.129297953
    sync_test.go:107: 
        	Error Trace:	/home/jgw/workspace/argo-cd/argocd-agent/test/e2e/sync_test.go:107
        	            				/usr/lib/golang/src/runtime/asm_amd64.s:1693
        	Error:      	Condition never satisfied
        	Test:       	TestSyncTestSuite/Test_SyncManaged
    fixture.go:107: Test ended at: 2026-01-29 12:53:27.862777208 -0500 EST m=+121.129533233
--- FAIL: TestSyncTestSuite (139.28s)
    --- FAIL: TestSyncTestSuite/Test_SyncManaged (139.24s)
FAIL
FAIL	github.com/argoproj-labs/argocd-agent/test/e2e	139.314s
FAIL
```

--- a/agent/connection.go
+++ b/agent/connection.go
@@ -169,8 +169,14 @@ func (a *Agent) handleStreamEvents() error {
 		return err
 	}
 
-	a.eventWriter = event.NewEventWriter(stream)
-	go a.eventWriter.SendWaitingEvents(a.context)
+	if a.eventWriter != nil {
+		// Reuse the existing event writer if it exists.
+		a.eventWriter.UpdateTarget(stream)
+	} else {
+		// Create a new event writer if it doesn't exist.
+		a.eventWriter = event.NewEventWriter(stream)
+		go a.eventWriter.SendWaitingEvents(a.context)
+	}
 
 	logCtx := log().WithFields(logrus.Fields{
 		logfields.Module:     "StreamEvent",

--- a/internal/event/event_writer.go
+++ b/internal/event/event_writer.go
@@ -15,6 +15,11 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 )
 
+const (
+	// maxEventRetries is the maximum number of times an event will be retried before giving up.
+	maxEventRetries = 5
+)
+
 type streamWriter interface {
 	Send(*eventstreamapi.Event) error
 	Context() context.Context
@@ -73,6 +78,16 @@ func (ew *EventWriter) UpdateTarget(target streamWriter) {
 	ew.mu.Lock()
 	defer ew.mu.Unlock()
 	ew.target = target
+
+	// Reset retry timers so events in sentEvents are retried immediately on the new connection.
+	// This is important for reconnection scenarios where the old connection died
+	// and ACKs will never arrive, so we want to give events a fresh chance on the new stream.
+	now := time.Now()
+	for _, msg := range ew.sentEvents {
+		msg.mu.Lock()
+		msg.retryAfter = &now
+		msg.mu.Unlock()
+	}
 }
 
 func (ew *EventWriter) Add(ev *cloudevents.Event) {
@@ -87,7 +102,7 @@ func (ew *EventWriter) Add(ev *cloudevents.Event) {
 	defer ew.mu.Unlock()
 
 	defaultBackoff := wait.Backoff{
-		Steps:    5,
+		Steps:    maxEventRetries,
 		Duration: 5 * time.Second,
 		Factor:   2.0,
 		Jitter:   0.1,
@@ -276,8 +291,7 @@ func (ew *EventWriter) retrySentEvent(resID string, sentMsg *eventMessage) {
 	})
 
 	// Check if we've exhausted retries
-	maxRetries := sentMsg.backoff.Steps
-	if sentMsg.retryCount >= maxRetries {
+	if sentMsg.retryCount >= maxEventRetries {
 		logCtx.Warnf("Event failed after %d retries, giving up to unblock queue", sentMsg.retryCount)
 		sentMsg.mu.Unlock()
 

--- a/internal/event/event_writer_test.go
+++ b/internal/event/event_writer_test.go
@@ -300,8 +300,7 @@ func TestEventWriter(t *testing.T) {
 		require.NotNil(t, sentMsg)
 
 		// Exhaust retries
-		maxRetries := sentMsg.backoff.Steps
-		for i := 0; i < maxRetries; i++ {
+		for i := 0; i <= maxEventRetries; i++ {
 			pastTime := time.Now().Add(-1 * time.Second)
 			sentMsg.retryAfter = &pastTime
 			evSender.retrySentEvent(resID, sentMsg)

